### PR TITLE
fix: prompt for new token when expired one is used

### PR DESCRIPTION
**Pull request by lucasmoura**
## Proposed Commit Message
fix: prompt for new token when expired one is used

When running ua fix we may need to install packages that depend on a UA service. Even if the user is already attached
to a UA subscription, this may not be enough, since the token used may be expired. If we encounter that situation, we would
need to ask the user to get a new token before proceeding with the package fixes. We are now adding the necessary code to allow us to handle that scenario.

Fixes: #1475

## Test Steps
Run the new unittests for this feature

## Desired commit type::

 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist:

 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: uaclient/security.py
Comment (blackboxsw) [inline uaclient/security.py:956]: Can we emit colorize commands for `ua refresh` here to represent the changes being performed on the system.
I filed a separate issue to make sure we do this in general for ua fix work https://github.com/canonical/ubuntu-advantage-client/issues/1477

**Review by blackboxsw — COMMENTED** [review]
Thanks Lucas, couple thoughts here on first pass.

File: uaclient/security.py
Comment (lucasmoura) [reply uaclient/security.py:956]: I though about doing that, however, I don't think we should surface that to the user. We are only running that command to check if the subscription is not expired, meaning that most of the times, it will probably not change the system at all.

But you think we can still output it, no problem

**Comment by lucasmoura** [comment]
@blackboxsw code updated

File: uaclient/security.py
Comment (blackboxsw) [reply uaclient/security.py:958]: I'm on the fence, but not fully convinced we need it, ua refresh does make a change if there are contract changes in general like changes in aptURL, packages required etc.  The only exposure I'd think about currently  is Ubuntu PRO images, once we release 26.2 to AWS/GCP/Azure, we will lock-step update contracts changing additionalPackages directives. I guess if a PRO image updated to 27.0 we might have a case where running a ua fix before reboot would attempt to apply those contract deltas and break on apt installs of ubuntu-fips metapkg due to current deb pinning. I suppose this is a small  corner-case and maybe negligible here.

File: uaclient/security.py
Comment (blackboxsw) [inline uaclient/security.py:956]: I have a pedantic request fo add a docstring here representing the ':return' intent for True vs False.

File: uaclient/security.py
Comment (blackboxsw) [inline uaclient/security.py:922]: What do you think: Is it worth sharing logic between _prompt_for_new_token and _prompt_for_attach?

Comment (lucasmoura) [reply uaclient/security.py:922]: I think the only part of the code we could reuse is this one:

```
print(status.colorize_commands([["ua", "attach", token]]))
return bool(
    0    
    == cli.action_attach(
        argparse.Namespace(token=token, auto_enable=True), cfg
    )    
)  
```

I think we can move it to a separate function. Let me do that

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Some valid integration test failures:

Some points that @mitechie should comment on:
 - I recently got guidance on some of the CVE messaging that if it's not a complete sentence, no need to follow with a period "." 
- [ ] I believe "See: ua status should not have a trailing stop because we expect possibly folks to cut-n-paste that command and the highlight might bring in the "."
- [ ] I believe "See: https:/ubuntu.com/advantage" doesn't need a trailing stop "." because it's not a sentence but a directive

Here's a diff of what I changed, dropping trailing stops from "See:  blah" and correcting a case where we also added a leading "." accidentally

diff: https://paste.ubuntu.com/p/vrWWWP9pSG/

success run: https://pastebin.canonical.com/p/SCDK47S37W/

File: uaclient/security.py
Comment (lucasmoura) [reply uaclient/security.py:958]: I think this is good point. Maybe we could have another way to check if the subscription is expired or not instead of running refresh. We could see if the contracts have an endpoint that allows us to do just that.

**Review by blackboxsw — APPROVED** [review]
Changes look good. +1 just waiting on CI before a merge

**Comment by blackboxsw** [comment]
CI known intermittent lxc failure that needs slightly better erroring from pycloudlib
````bash
        File "/jenkins/servers/private/workspace/_ubuntu-advantage-client_PR-1480@3/.tox/behave-lxd-14.04/lib/python3.6/site-packages/pycloudlib/lxd/instance.py", line 88, in ip
      ip_address = result.split()[0]
  IndexError: list index out of range
```


Already a pycloudlib issue for this  https://github.com/canonical/pycloudlib/issues/129

**Pull request merged**
